python/pyspark/ml/functions.py

```diff
@@ -106,6 +152,597 @@ def array_to_vector(col: Column) -> Column:
+                "columns: expected {}, got: {}".format(fieldNames, predNames)
+            )
+
+        return pd.DataFrame(preds)
+    elif isinstance(return_type, ArrayType):
+        if isinstance(preds, np.ndarray):
+            if len(preds) != num_input_rows:
+                raise ValueError("Prediction results must have same length as input data.")
+            if len(preds.shape) != 2:
+                raise ValueError("Prediction results for ArrayType must be two-dimensional.")
+        else:
+            raise ValueError("Prediction results for ArrayType must be an ndarray.")
+
+        return pd.Series(list(preds))
+    elif isinstance(return_type, supported_scalar_types):
+        preds_array: np.ndarray = preds  # type: ignore
+        if len(preds_array) != num_input_rows:
+            raise ValueError("Prediction results must have same length as input data.")
+        if not (
+            (len(preds_array.shape) == 2 and preds_array.shape[1] == 1)
+            or len(preds_array.shape) == 1
+        ):
+            raise ValueError("Invalid shape for scalar prediction result.")
+
+        return pd.Series(np.squeeze(preds))  # type: ignore
+    else:
+        raise ValueError("Unsupported return type")
+
+
+def predict_batch_udf(
+    make_predict_fn: Callable[
+        [],
+        PredictFunction,
+    ],
+    *,
+    return_type: DataType,
+    batch_size: int,
+    input_tensor_shapes: List[List[int] | None] | Mapping[int, List[int]] | None = None,
+) -> UserDefinedFunctionLike:
+    """Given a function which loads a model and makes batch predictions, returns a Pandas UDF for
+    inferencing over that model.
+
+    The returned UDF does the following on each DataFrame partition:
+    - calls `make_predict_fn` to load the model and cache its `predict_fn`.
+    - batches the input records as numpy arrays and invokes `predict_fn` on each batch.
+
+    This assumes that the `make_predict_fn` encapsulates all of the necessary dependencies for
+    running the model or the Spark executor environment already satisfies all runtime requirements.
+
+    For the conversion of Spark DataFrame to numpy arrays, there is a one-to-one mapping between the
+    input arguments of the `predict_fn` (returned by the `make_predict_fn`) and the input columns to
+    the UDF (returned by the `predict_batch_udf`) at runtime.  Each input column will be converted
+    as follows:
+    - scalar column -> np.ndarray
+    - tensor column + tensor shape -> np.ndarray
+
+    Note that tensor columns in the Spark DataFrame must be represented as a flattened 1-D array,
+    and multiple scalar columns can be combined into a single tensor column using standard PySpark
+    SQL functions like `array()` or `struct()`.
+
+    Example (tensor column):
+
+    Input DataFrame has a single column with a flattened tensor value, represented as an array of
+    float.
+    ```
+    from pyspark.ml.functions import predict_batch_udf
+
+    def make_predict_fn():
```

Comment:
nit: it would be nice to rename it according to what it does, e.g., `make_mnist_fn`